src/D2L.CodeStyle.Analyzers/ParallelizableTestsAnalyzer.cs

```diff
@@ -0,0 +1,130 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers {
+    [DiagnosticAnalyzer( LanguageNames.CSharp )]
+    public class ParallelizableTestsAnalyzer : DiagnosticAnalyzer {
+        public const string DiagnosticId = "D2L0001";
+        private const string Category = "Safety";
+
+        private const string Title = "Ensure test is parallelizable.";
+        private const string Description = "Classes with side-effects should not be used in non-Isolated tests.";
+        internal const string MessageFormat = "'{0}' is not safe for use outside of isolated tests. Consider marking this test as isolated, or use `{1}` instead.";
+
+        private static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+            DiagnosticId, Title, MessageFormat, Category,
+            DiagnosticSeverity.Error,
+            isEnabledByDefault: true,
+            description: Description
+        );
+
+        private static readonly ImmutableDictionary<string, string> DefaultOffendingTypeSuggestions = new Dictionary<string, string> {
+            {"D2L.LP.TestFramework.Configuration.FeatureToggling.Domain.ITestFeatureToggler", "D2L.LP.Configuration.FeatureToggling.Domain.TestFeatureToggleFactory" },
+            {"D2L.LP.TestFramework.Configuration.FeatureToggling.Domain.TestFeatureTogglerExtensions", "D2L.LP.Configuration.FeatureToggling.Domain.TestFeatureToggleFactory" }
+        }.ToImmutableDictionary();
+
+        private readonly ImmutableDictionary<string, string> m_offendingTypeSuggestions;
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create( Rule );
+
+        /// <summary>
+        /// Default constructor
+        /// </summary>
+        public ParallelizableTestsAnalyzer() : this( DefaultOffendingTypeSuggestions ) { }
+
+        internal ParallelizableTestsAnalyzer( ImmutableDictionary<string, string> ofendingTypeSuggestionsOverride ) {
+            m_offendingTypeSuggestions = ofendingTypeSuggestionsOverride;
+        }
+
+        public override void Initialize( AnalysisContext context ) {
+            context.RegisterCompilationStartAction( RegisterIfTestAssembly );
+        }
+
+        private void RegisterIfTestAssembly( CompilationStartAnalysisContext compilation ) {
+            var references = compilation.Compilation.ReferencedAssemblyNames;
+            if( !references.Any( r => r.Name.ToUpper().Contains( "NUNIT" ) ) ) {

```

Comment:
Nice. Worth validating this is legit in the monolith. I have vague memories that we have some prod assemblies dep'ing on this (super wrong)